src/compiler/binder.ts

```diff
@@ -2722,8 +2721,9 @@ namespace ts {
         return transformFlags & ~TransformFlags.ArrayLiteralOrCallOrNewExcludes;
     }
 
-    function isSuperOrSuperProperty(node: Node, kind: SyntaxKind) {
-        switch (kind) {
+    function isSuperOrSuperProperty(node: Node) {
+        node = skipOuterExpressions(node);

```

Comment:
@rbuckton According to the commit log, `kind` was being passed in here to avoid some kind of polymorphism. If that's the case; would it make sense to continue having `isSuperOrSuperProperty` take the `kind` and `node` directly, and move the call to `skipOuterExpressions` into `computeCallExpression`, or is what I have here (which is more idiomatic considering what we do elsewhere) fine?

Reply:
We would want to have the call to `skipOuterExpressions` in `computeCallExpression`.

Reply:
This was an optimization introduced after reviewing deoptimization data via IRHydra. By only digging into `node` when `kind` indicates a property or element access, we avoid a number of soft-deoptimization hits from looking at both `node.kind` and `node.expression` in the same function.

The most efficient approach which doesn't require traversing the subtree tree again would be to have a `TransformFlags.ContainsSuperOrSuperProperty`. We would then add the flag for `SuperKeyword` and for a property or element access whose `expression` is `SuperKeyword` and reset the flag for all nodes *except* `ParenthesizedExpression`, `AsExpression`, `TypeAssertionExpression`, or `PartiallyEmittedExpression` so that it doesn't leak out of the context where it needs to matter.

Then we only need to interrogate `subtreeFlags` for the answer, which avoids polymorphism and walking the subtree.

If we are running out of `TransformFlags` then it doesn't matter where we put it, since the cost of deoptimization is probably negligible compared to the cost of `skipOuterExpressions` (which likely ends up deoptimized in any case since it is highly polymorphic).

Reply:
We have two unused bits left in `TransformFlags`; `27` and `28`, so we can do it. But even as is the next new emit feature that seriously needs transform flags (like, say, proper lexical `arguments`) is already in for a bad time what with there not being enough flags. but I suppose that's a future problem.